### PR TITLE
fix(normaliseUrlDomain): allow custom url scheme

### PR DESCRIPTION
**Pull request by zaosoula**
## Summary of change
Refactor the code of normaliseUrlDomainOrThrowError to accept schemes other than http://, https://, supertokens://
(A few sentences about this PR)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] If new thirdparty provider is added,
    -   [ ] update switch statement in `recipe/thirdparty/providers/configUtils.ts` file, `createProvider` function.
    -   [ ] add an icon on the user management dashboard.
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [ ] Setup the testing env to ensure that no breaking change has been introduced

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit e6c547703bebc6d7c6e378a78d9a394abe92a3ea.  | 
|--------|--------|

### Summary:
The PR modifies the `normaliseURLDomainOrThrowError` function to accept URL schemes other than `http://`, `https://`, and `supertokens://`, by adding a default protocol if none is provided, and returning the URL string with the appropriate protocol and host.

**Key points**:
- Modified `normaliseURLDomainOrThrowError` function in `/lib/build/normalisedURLDomain.js` and `/lib/ts/normalisedURLDomain.ts` files.
- The function now checks if the input starts with a protocol and adds a default one if not.
- If the input starts with `localhost` or is an IP address, `http://` is added. Otherwise, `https://` is added.
- The function tries to create a new URL object with the input. If successful, it returns the URL string with the appropriate protocol and host. If the URL object creation fails, it throws an error.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by zaosoula** [comment]
@rishabhpoddar Here is the pr, unfortunatly I'm unable to launch the test env for now, so if you can give it a try

**Comment by rishabhpoddar** [comment]
@zaosoula i'll probably try it out later today or sometime this week. In the meantime, you can use your forked version of the sdk instead of the one from npm in your app and see if it works well. Also, you may have to make a similar change to the frontend sdk as well cause that too has the same normalised domain class

File: lib/ts/normalisedURLDomain.ts [hunk 1]
```diff
@@ -30,48 +30,33 @@ export default class NormalisedURLDomain {
 function normaliseURLDomainOrThrowError(input: string, ignoreProtocol = false): string {
```
Comment (rishabhpoddar) [inline lib/ts/normalisedURLDomain.ts:30]:
```suggestion
function normaliseURLDomainOrThrowError(input: string): string {
```

Comment (rishabhpoddar) [reply lib/ts/normalisedURLDomain.ts:30]: Not needed since we are not recursing anymore

File: lib/ts/normalisedURLDomain.ts [hunk 2]
```diff
@@ -30,48 +30,33 @@ export default class NormalisedURLDomain {
...
+        } else {
+            input = "https://" + input;
         }
-        let urlObj = new URL(input);
+    }
+
+    try {
+        const urlObj = new URL(input);
+
         if (ignoreProtocol) {
```
Comment (rishabhpoddar) [inline lib/ts/normalisedURLDomain.ts:50]: this if block will never be true so can be removed.

File: lib/ts/normalisedURLDomain.ts [hunk 3]
```diff
@@ -30,48 +30,33 @@ export default class NormalisedURLDomain {
...
-    try {
-        if (!input.startsWith("http://") && !input.startsWith("https://") && !input.startsWith("supertokens://")) {
-            throw new Error("converting to proper URL");
+    // if the input starts with a . (eg: .domain.tld)
+    if (input.indexOf(".") === 0) {
+        input = input.substring(1);
+    }
+
+    // if the input dosen't start with a protocol add a default one;
+    if (input.match(/^[^:]:\/\//)) {
```
Comment (rishabhpoddar) [inline lib/ts/normalisedURLDomain.ts:39]: im not sure if this regex is correct. Cause if i do `example.com".match(/^[^:]:\/\//)`, it returns null. But we want it to come in the if block in this case. Am i missing something?

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Comment by rishabhpoddar** [comment]
Hey @ellipsis, give me a code review

**Comment by zaosoula** [comment]
Hey https://github.com/ellipsis, give me a code review

**Pull request closed without merging**